pkg/dubbo/parameter.go

```diff
@@ -0,0 +1,239 @@
+/*
+ * Copyright 2023 CloudWeGo Authors
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * This source file has been replicated from the original dubbo-go project
+ * repository, and we extend our sincere appreciation to the dubbo-go
+ * development team for their valuable contribution.
+ */
+
+package dubbo
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+	"sync"
+	"time"
+
+	hessian "github.com/apache/dubbo-go-hessian2"
+)
+
+var cache = new(typesCache)
+
+// typesCache maintains a cache from type of data(reflect.Type) to Types string used by Hessian2.
+type typesCache struct {
+	group    Group
+	typesMap sync.Map
+}
+
+// getByData returns the Types string of given data.
+// It reads embedded sync.Map firstly. If cache misses, using singleFlight to process reflection and getParamsTypeList.
+func (tc *typesCache) getByData(data interface{}) (string, error) {
+	val := reflect.ValueOf(data)
+	typ := val.Type()
+	typesRaw, ok := tc.typesMap.Load(typ)
+	if ok {
+		return typesRaw.(string), nil
+	}
+
+	typesRaw, err, _ := tc.group.Do(typ, func() (interface{}, error) {
+		elem := val.Elem()
+		numField := elem.NumField()
+		fields := make([]interface{}, numField)
+		for i := 0; i < numField; i++ {
+			fields[i] = elem.Field(i).Interface()
+		}
+
+		types, err := getParamsTypeList(fields)
+		if err != nil {
+			return "", err
+		}
+
+		return types, nil
+	})
+	if err != nil {
+		return "", err
+	}
+	tc.typesMap.Store(typ, typesRaw)
```

Comment:
1. The `Store` operation should be called within the `Do` function.
2. if getParamsTypeList won't return an error in any case, we can safely ignore it.